src/make_watertight/Gen.cpp

```diff
@@ -0,0 +1,2117 @@
+  assert(moab::MB_SUCCESS == result);
+
+  // project the t_joint between the endpts
+  // http://en.wikipedia.org/wiki/Vector_projection
+  moab::CartVect a = coords[0] - coords[1];
+  moab::CartVect b = coords[2] - coords[1];
+  parameter    = (a % b) / (b % b);
+  moab::CartVect c = parameter * b;
+  projected_coords = c     + coords[1];
+  return moab::MB_SUCCESS;
+}
+
+moab::ErrorCode Gen::point_line_projection(const moab::EntityHandle line_pt1,
+                                           const moab::EntityHandle line_pt2,
+                                           const moab::EntityHandle pt0,
+                                           double& dist_along_edge) {
+
+  moab::ErrorCode result;
+  moab::CartVect coords[3];
+  result = MBI()->get_coords(&line_pt1, 1, coords[1].array());
+  assert(moab::MB_SUCCESS == result);
+  result = MBI()->get_coords(&line_pt2, 1, coords[2].array());
+  assert(moab::MB_SUCCESS == result);
+  result = MBI()->get_coords(&pt0, 1, coords[0].array());
+  assert(moab::MB_SUCCESS == result);
+
+  // project the t_joint between the endpts
+  // http://en.wikipedia.org/wiki/Vector_projection
+  moab::CartVect a = coords[0] - coords[1];
+  moab::CartVect b = coords[2] - coords[1];
+  dist_along_edge = a % b / b.length();
+  return moab::MB_SUCCESS;
+}
+
+double Gen::area2(const moab::EntityHandle pt_a, const moab::EntityHandle pt_b,
+                  const moab::EntityHandle pt_c, const moab::CartVect plane_normal) {
+  moab::ErrorCode result;
+  moab::CartVect a, b, c;
+  result = MBI()->get_coords(&pt_a, 1, a.array());
+  assert(moab::MB_SUCCESS == result);
+  result = MBI()->get_coords(&pt_b, 1, b.array());
+  assert(moab::MB_SUCCESS == result);
+  result = MBI()->get_coords(&pt_c, 1, c.array());
+  assert(moab::MB_SUCCESS == result);
+  moab::CartVect d = b - a;
+  moab::CartVect e = c - a;
+  // project onto a plane defined by the plane's normal vector
+  return (d * e) % plane_normal;
+}
+
+// Is point c to the left of line ab?
+bool Gen::left(const moab::EntityHandle a, const moab::EntityHandle b,
+               const moab::EntityHandle c, const moab::CartVect n) {
+  double area_2 = area2(a, b, c, n);
+  if (area_2 > 0)
+    return true;
+  else
+    return false;
+}
+
+// Is point c to the left of line ab or collinear?
+bool Gen::left_on(const moab::EntityHandle a, const moab::EntityHandle b,
+                  const moab::EntityHandle c, const moab::CartVect n) {
+  double area_2 = area2(a, b, c, n);
+  if (area_2 >= 0)
+    return true;
+  else
+    return false;
+}
+
+// Are pts a,b,c collinear?
+bool Gen::collinear(const moab::EntityHandle a, const moab::EntityHandle b,
+                    const moab::EntityHandle c, const moab::CartVect n) {
+  double area_2 = area2(a, b, c, n);
+  if (area_2 == 0)
+    return true;
+  else
+    return false;
+}
+
+// Exclusive or: T iff exactly one argument is true
+bool logical_xor(const bool x, const bool y) {
+  return (x || y) && !(x && y);
```

Comment:
Readability: You can use the `!=` operator with booleans to do an xor [1].

[1] https://stackoverflow.com/a/1596681/1015951